drivers/power/lt7170/iio_lt7170.c

```diff
@@ -0,0 +1,1039 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include "no_os_alloc.h"
+#include "no_os_error.h"
+#include "no_os_units.h"
+#include "no_os_util.h"
+
+#include "lt7170.h"
+#include "iio_lt7170.h"
+
+static const char *const lt7170_enable_avail[2] = {
+	"Disabled", "Enabled"
+};
+
+enum lt7170_iio_enable_type {
+	LT7170_IIO_OUTPUT_ENABLE,
+	LT7170_IIO_PULSE_ENABLE,
+	LT7170_IIO_SYNC_ENABLE,
+};
+
+enum lt7170_iio_pwm_params {
+	LT7170_IIO_PWM_FREQUENCY,
+	LT7170_IIO_PWM_PHASE,
+};
+
+enum lt7170_iio_chan_type {
+	LT7170_IIO_VOUT_CHAN,
+	LT7170_IIO_VIN_CHAN,
+	LT7170_IIO_IOUT_CHAN,
+	LT7170_IIO_TEMP_CHAN,
+	LT7170_IIO_VCC_CHAN,
+};
+
+static struct iio_device lt7170_iio_dev;
+
+/**
+ * @brief Read register value.
+ * @param dev     - The iio device structure.
+ * @param reg	  - Register to read.
+ * @param readval - Read value.
+ * @return ret    - Result of the reading procedure.
+ * 		    In case of success, the size of the read data is returned.
+*/
+static int32_t lt7170_iio_reg_read(void *dev, uint32_t reg, uint32_t *readval)
+{
+	struct lt7170_iio_desc *iio_lt7170 = dev;
+	struct lt7170_dev *lt7170 = iio_lt7170->lt7170_dev;
+	int ret;
+	uint8_t block[4] = {0};
+
+	switch (reg) {
+	case LT7170_PAGE:
+	case LT7170_OPERATION:
+	case LT7170_ON_OFF_CONFIG:
+	case LT7170_WRITE_PROTECT:
+	case LT7170_CAPABILITY:
+	case LT7170_VOUT_MODE:
+	case LT7170_VOUT_OV_FAULT_RESPONSE:
+	case LT7170_VOUT_UV_FAULT_RESPONSE:
+	case LT7170_IOUT_OC_FAULT_RESPONSE:
+	case LT7170_OT_FAULT_RESPONSE:
+	case LT7170_VIN_OV_FAULT_RESPONSE:
+	case LT7170_TON_MAX_FAULT_RESPONSE:
+	case LT7170_STATUS_BYTE:
+	case LT7170_STATUS_VOUT:
+	case LT7170_STATUS_IOUT:
+	case LT7170_STATUS_INPUT:
+	case LT7170_STATUS_TEMPERATURE:
+	case LT7170_STATUS_CML:
+	case LT7170_STATUS_MFR_SPECIFIC:
+	case LT7170_REVISION:
+	case LT7170_MFR_FAULT_RESPONSE:
+	case LT7170_MFR_ADC_CONTROL:
+	case LT7170_MFR_COMMON:
+	case LT7170_MFR_CHANNEL_STATE:
+	case LT7170_MFR_SYNC_CONFIG:
+	case LT7170_MFR_RAIL_ADDRESS:
+	case LT7170_MFR_DISABLE_OUTPUT:
+		return lt7170_read_byte(lt7170, (uint8_t)reg,
+					(uint8_t *)readval);
```

Comment:
You will have to use a separate `uint8_t` variable for this. Currently, the result will be different on big endian architectures

Reply:
Hi, I apologize I'm a bit confused. Could you kindly elaborate on this? I referred to this implementation on that line of code:  [LT7182S_IIO](https://github.com/analogdevicesinc/no-OS/blob/main/drivers/power/lt7182s/iio_lt7182s.c#L123)

Reply:
The function call should be done as following:

```
uint8_t reg_val;
lt7170_read_byte(lt7170, reg, &readval);
*read_val = reg_val;
```

Since your `lt7170_read_byte()` function takes a parameter that's a pointer to an 1 byte unsigned int, it will end up setting the first byte at the address `readval`. Depending on the microcontroller endianess, the actual result will be interpreted as `0xFE000000` or `0x000000FE`, so the values will be different.

Another issue is that by setting only 1 byte, you don't clear any non 0 values in that memory region, so the result will be wrong. For example:

uint32_t reg_val; /* Not initialized, so its value is set to whatever was on the stack */
lt7170_iio_reg_read(..., &reg_val);

At this point, only the first byte of `reg_val` was set, so your result will be wrong.